component/espejote-templates/upgrade-notification.jsonnet

```diff
@@ -0,0 +1,72 @@
+local esp = import 'espejote.libsonnet';
+local config = import 'minor-upgrade-notification/config.json';
+local dst = import 'minor-upgrade-notification/dst.json';
+
+// Finds the next (i.e. "oldest") clusterversion overlay that sets a channel
+// higher than the current OpenShift minor version.
+local nextChannelOverlay =
+  local overlays = esp.context().clusterversion_appuio[0].spec.overlays;
+  local current_minor = std.split(esp.context().clusterversion_ocp[0].spec.desiredUpdate.version, '.')[1];
+  local pendingOverlays = [
+    [overlay.from, overlay.overlay.spec.channel]
+    for overlay in overlays
+    if std.objectHas(overlay.overlay.spec, 'channel')
+      && std.split(overlay.overlay.spec.channel, '.')[1] > current_minor
+  ];
+  if std.length(pendingOverlays) > 0 then
+    local next = std.sort(pendingOverlays)[0];
+    {date: next[0], channel: next[1]};
+
+// Finds the first upgrade window after the overlay date. Upgradeconfigs only
+// display the next 10 maintenance windows, so it's not guaranteed to find one
+// if the overlay is far in the future.
+local upgradeWindow = if nextChannelOverlay != null then
+  local possible = [t.time for t in esp.context().upgradeconfig[0].status.nextPossibleSchedules if t.time > nextChannelOverlay.date ];
+  if std.length(possible) > 0 then possible[0];
+
+// Custom format datetime and calculate Europe/Zurich time based on dates where
+// daylight saving times change.
+local formatDate(date) =
+  local lastChange = std.reverse([t for t in std.objectFields(dst) if t < date])[0];
+  local UTCOffset = dst[lastChange];
+  local YMD = std.split(date, 'T')[0];
+  local Y = std.split(YMD, '-')[0];
+  local M = std.split(YMD, '-')[1];
+  local D = std.split(YMD, '-')[2];
+  local hms = std.split(date, 'T')[1];
+  local h = std.toString(std.parseInt(std.split(hms, ':')[0]) + UTCOffset);
+  local m = std.split(hms, ':')[1];
+  '%s.%s.%s %s:%s' % [D, M, Y, h, m];
+
+local replacementValues = {
+  '$OVERLAY_DATE': formatDate(nextChannelOverlay.date),
+  '$NEXT_MAINTENANCE': formatDate(upgradeWindow),
+  '$OVERLAY_CHANNEL': nextChannelOverlay.channel,
+  '$OVERLAY_VERSION': std.split(nextChannelOverlay.channel, '-')[1],
+  '$OVERLAY_VERSION_MINOR': std.split(nextChannelOverlay.channel, '.')[1],
+};
+
+local makeConsoleNotification(name, args, repl) =
+  local replace(t, v) = std.strReplace(t, v, std.get(repl, v));
+  local text = std.foldl(replace, std.objectFields(repl), args.text);
```

Comment:
without fully testing it, I believe that the current implementation (given `nextChannelOverlay.channel="4.18"`) would replace `$OVERLAY_VERSION_MINOR` with `4.18_MINOR` because you replace `$OVERLAY_VERSION` before `$OVERLAY_VERSION_MINOR` and `strReplace()` doesn't actually understand that you're emulating shell variables.

Also this implementation (in contrast to real shell variables) doesn't recognize `${OVERLAY_VERSION}` as referencing the same replacement.

Reply:
https://github.com/appuio/component-openshift4-console/commit/583d9591bf0c22bf9a4fcb1cd00591edc5812c54